pkg/ruler/notifier.go

```diff
@@ -74,67 +73,74 @@ func (rn *rulerNotifier) stop() {
-			Names:           []string{u.Host},
-			RefreshInterval: model.Duration(rulerConfig.AlertmanagerRefreshInterval),
-			Type:            "SRV",
-			Port:            0, // Ignored, because of SRV.
+	for _, h := range rulerConfig.AlertmanagerURL {
+		url, err := url.Parse(h)
+		if err != nil {
+			return nil, err
 		}
-		sdConfig = sd_config.ServiceDiscoveryConfig{
-			DNSSDConfigs: []*dns.SDConfig{&dnsSDConfig},
-		}
-	} else {
-		sdConfig = sd_config.ServiceDiscoveryConfig{
-			StaticConfigs: []*targetgroup.Group{
-				{
-					Targets: []model.LabelSet{
-						{
-							model.AddressLabel: model.LabelValue(u.Host),
-						},
-					},
-				},
-			},
+
+		if url.String() == "" {
+			continue
 		}
+
+		validURLs = append(validURLs, url)
 	}
 
-	amConfig := &config.AlertmanagerConfig{
-		APIVersion:             config.AlertmanagerAPIVersionV1,
-		Scheme:                 u.Scheme,
-		PathPrefix:             u.Path,
-		Timeout:                model.Duration(rulerConfig.NotificationTimeout),
-		ServiceDiscoveryConfig: sdConfig,
+	if len(validURLs) == 0 {
+		return &config.Config{}, nil
 	}
 
+	apiVersion := config.AlertmanagerAPIVersionV1
 	if rulerConfig.AlertmanangerEnableV2API {
-		amConfig.APIVersion = config.AlertmanagerAPIVersionV2
+		apiVersion = config.AlertmanagerAPIVersionV2
 	}
 
-	promConfig := &config.Config{
-		AlertingConfig: config.AlertingConfig{
-			AlertmanagerConfigs: []*config.AlertmanagerConfig{amConfig},
-		},
-	}
+	amConfigs := make([]*config.AlertmanagerConfig, 0, len(validURLs))
+	for _, url := range validURLs {
+
+		var sdConfig sd_config.ServiceDiscoveryConfig
+		if rulerConfig.AlertmanagerDiscovery {
+			sdConfig.DNSSDConfigs = []*dns.SDConfig{{
+				Names:           []string{url.Host},
+				RefreshInterval: model.Duration(rulerConfig.AlertmanagerRefreshInterval),
+				Type:            "SRV",
+				Port:            0, // Ignored, because of SRV.
+			}}
+		} else {
+			sdConfig.StaticConfigs = []*targetgroup.Group{{
+				Targets: []model.LabelSet{{model.AddressLabel: model.LabelValue(url.Host)}},
+			}}
+		}
 
-	if u.User != nil {
-		amConfig.HTTPClientConfig = config_util.HTTPClientConfig{
-			BasicAuth: &config_util.BasicAuth{
-				Username: u.User.Username(),
-			},
+		amConfig := &config.AlertmanagerConfig{
+			APIVersion:             apiVersion,
+			Scheme:                 url.Scheme,
+			PathPrefix:             url.Path,
+			Timeout:                model.Duration(rulerConfig.NotificationTimeout),
+			ServiceDiscoveryConfig: sdConfig,
 		}
 
-		if password, isSet := u.User.Password(); isSet {
-			amConfig.HTTPClientConfig.BasicAuth.Password = config_util.Secret(password)
+		if url.User != nil {
+			amConfig.HTTPClientConfig = config_util.HTTPClientConfig{
+				BasicAuth: &config_util.BasicAuth{
+					Username: url.User.Username(),
+				},
+			}
+
+			if password, isSet := url.User.Password(); isSet {
+				amConfig.HTTPClientConfig.BasicAuth.Password = config_util.Secret(password)
+			}
 		}
+
```

Comment:
Can we abstract this out into a function like `amConfigFromURL`? Its easier to read that way.